### ISSUE TITLE
"_" patterns are strange

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#261]
The following code is not even unsafe, so it certainly cannot be UB:
```rust
fn foo(ptr: *const bool) {
    let _ = *ptr;
}

foo(&3u8 as *const u8 as *const bool);
```
This indicates that the pointer does not really get dereferenced, so the pointed-to data is never interpreted at type `bool` and thus must not be valid. I am not sure I like this; `*ptr` (as a value expression!) to me seems like it should conceptually perform the load even if the result is later discarded.

Something similar happens here:
```rust
fn bar(ptr: *const (i8, bool)) { unsafe {
    let (x, _) = *ptr;
} }
```
This compiles effectively to `let x = (*ptr).1`, so again whatever data is in the second field does not matter. (Note that `*ptr` here occurs only as a place expression, not as a value expression, so it makes sense that in `(*ptr).1` we do not require the second field to be valid.)

I think this behavior of "_" can be explained by saying that it suppresses the place-to-value coercion, and it is smart enough to even do this in cases like `(x, _)` where the coercion is *partially* suppressed. But I am not sure I like this semantics, it seems rather counter-intuitive to me. It also makes patterns much harder to explain. And it leads to strange special cases such as https://github.com/rust-lang/rust/issues/79735 (I cannot tell if this is some compiler author also being confused by "_", or if the intention is that `!` is special in that even creating the *place* is already UB, or something else).

**Comment by RalfJung** [comment]
@Nadrieril points out that

> Right now you are allowed to write let (y, _) = x if we moved out x.1 already. The corresponding match is not allowed tho

**Comment by RalfJung** [comment]
The corresponding case for partial initialization can currently not be probed since rustc no longer permits writing to individual fields of a not-yet-initialized struct:
```rust
fn main() {
    let x: (i32, bool);
    x.0 = 5; // this already errors here...
    let (y, _) = x; // ... so we cannot test if this would be accepted
}
```
If this code was accepted, that would be further evidence for `_` patterns inhibiting place-to-value coercions.

**Comment by RustyYato** [comment]
Like this?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3d2143d5f263fb1a6d324c65feb0dc31

```rust
fn main() {
    let x = (0, String::new());
    drop(x.1);
    let (x, _) = x;
    // let (x, y) = x; // doesn't compile
}
```

**Comment by RalfJung** [comment]
`drop` doesn't change the content of the location though, the value still adheres to the validity invariant (and it has to, otherwise `ManuallyDrop` would be unsound).

**Comment by scottmcm** [comment]
>The following code is not even unsafe, so it certainly cannot be UB:

[It _was_ `unsafe` back in <=1.21, though.](https://rust.godbolt.org/z/YvqEdj)  Is it not-unsafe on purpose, or by accident in the MIR unsafety check?

**Comment by tmiasko** [comment]
The MIR for `let _ = *p;` is empty, so it could well be accidental.

**Comment by elichai** [comment]
This is IMHO even more confusing because `let _ = a; ` doesn't even drop `a`, so it feels to me like `let _ = $expr` doesn't discard the expression so the read should actually read.
(on the other hand you could argue that this is because $expr is just removed from MIR, and then it won't drop + it won't read, but that feels less intuitive for someone who doesn't think about IR when writing code)

**Comment by RalfJung** [comment]
> let _ = a; doesn't even drop a

Well, it drops `a` sometimes...

```rust
fn mark() {}

fn test1(x: Vec<i32>) {
    let _ = x; // nothing happens here
    mark();
}

fn test2(x: Vec<i32>) {
    let _ = vec![1]; // this is dropped here
    mark();
}
```

**Comment by bjorn3** [comment]
In `test2`, it isn't the `_` that drops it. It is the fact that the temporary `vec![1]` goes out of scope that drops it.

**Comment by RalfJung** [comment]
Well, but the `_` is what makes it "go out of scope", so I don't think one can entirely separate these two things.

**Comment by gThorondorsen** [comment]
> One could just as well argue that `let _ = x` in `test1` makes "`x` go out of scope".

Currently, [it does not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=00963583dcce6bee88e34832f0f8b166). But it does in [this very similar test](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9b0a77d909ae01cf756099fb673ff4d8) using `let _ = {x};` instead.

**Comment by RalfJung** [comment]
> Currently, it does not.

Yeah, it doesn't, but by analogy with `let _ = vec![1];`, maybe it should.

The current situation seems mostly consistent (except for https://github.com/rust-lang/rust/issues/79735), but it's strange because behavior heavily depends on whether the expression on the right evaluate to a temporary or an existing place.

**Comment by Nadrieril** [comment]
I guess it's for consistency with the fact that `let (y, _) = x` only moves out of `x.0`. In that context we don't want `_` to move anything out

**Comment by RalfJung** [comment]
@scottmcm also points out that there seems to be a difference between the *statements* `let _ = *ptr;` and `*ptr;`, which also seems odd at best.

**Comment by camelid** [comment]
> This compiles effectively to `let x = (*ptr).1`, so again whatever data is in the second field does not matter. (Note that `*ptr` here occurs only as a place expression, not as a value expression, so it makes sense that in `(*ptr).1` we do not require the second field to be valid.)

@RalfJung I think you mean `(*ptr).0` here?

**Comment by RalfJung** [comment]
Ah, yes... I am switching too often between languages that index tuples starting at 0 vs 1. ;)

**Comment by RalfJung** [comment]
@pnkfelix points out that adding a type annotations make a difference for the operational behavior of `_` patterns... https://github.com/rust-lang/rust/issues/80059#issuecomment-756869717.

**Comment by RalfJung** [comment]
Also Cc https://github.com/rust-lang/miri/issues/2360, we probably want to consider some of the code around `_` UB that is currently accepted by Miri.

**Comment by RalfJung** [comment]
With https://github.com/rust-lang/rust/pull/104844 and https://github.com/rust-lang/rust/pull/103208 having landed, the issue is mostly resolved now: `_` patterns construct a place but never load from it. That means `let _ = place;` is UB if and only if `addr_of!(place)` is UB, and same for the other ways of using `_` patterns.

There's a bug where this is not quite true for the `!` type: https://github.com/rust-lang/rust/issues/117288. But I think that's just a bug, the semantic questions are all clear.

Do we have official docs anywhere on the semantics of `_` patterns? That seems to be the last missing bit.
